### DISCUSSION TITLE
v5.0.0

### DISCUSSION THREAD

**Discussion by novemberborn** [avajs/ava#3102] [General]
Despite the major version bump this is a relatively minor release. Node.js 12 is no longer supported. The type definitions are now distributed to be compatible with TypeScript 4.7 or newer.

It's available as a pre-release under the `next` tag for the next few weeks:

```console
npm install --save-dev ava@next
```

## Breaking Changes

* AVA now requires Node.js 14.19, 16.15 or 18. https://github.com/avajs/ava/pull/3043
* Type definition distribution is no longer compatible with TypeScript versions before 4.7. https://github.com/avajs/ava/pull/3041
* We're now exporting types rather than interfaces. https://github.com/avajs/ava/pull/3096

## Improvements

* Detect `process.exit()` called from tests by @gibson042 in https://github.com/avajs/ava/pull/3080
* Improve test result accessibility by @gibson042 in https://github.com/avajs/ava/pull/3082
* Parse for line numbers with latest supported ECMA version by @novemberborn in https://github.com/avajs/ava/pull/3063
* Work around `os.cpus()` returning an empty array on unsupported platforms by @catink123 in https://github.com/avajs/ava/pull/3095
* Count timeouts separately from the timed out tests by @novemberborn in https://github.com/avajs/ava/pull/3099

## New Contributors

* @Prinzhorn made their first contribution in https://github.com/avajs/ava/pull/3073
* @catink123 made their first contribution in https://github.com/avajs/ava/pull/3095

**Full Changelog**: https://github.com/avajs/ava/compare/v4.3.3...v5.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/avajs/ava/releases/tag/v5.0.0'>v5.0.0</a>.</em>

**Comment by novemberborn** [comment]
Please let me know how this is working for you, especially the type definition changes. If it's all good I'll promote to `latest` on npm.

**Reply by catink123** [reply]
It seems that importing the library in the test file throws an error with TypeScript 4.7 (as required in 5.0.0). It says "Cannot find module 'ava' or its corresponding type declarations". Importing from "ava/entrypoints/main" obviously isn't allowed as it's not exported.

Important package.json entries: 
```json
  "devDependencies": {
    "@ava/typescript": "^3.0.1",
    "@types/node": "^18.7.16",
    "ava": "^5.0.0",
    "typescript": "^4.7.4"
  },
  "ava": {
    "typescript": {
      "rewritePaths": {
        "src/": "dist/"
      },
      "compile": "tsc"
    }
  },
```

**Reply by novemberborn** [reply]
Could you share the command and full output?

**Reply by catink123** [reply]
Here it is:

```
$ yarn test
yarn run v1.22.19
$ ava src/test.ts

  ✘ Internal error
  Error: Command failed with exit code 2: tsc --incremental
  src/test.ts(1,49): error TS2307: Cannot find module 'ava' or its corresponding type declarations.
  Error: Command failed with exit code 2: tsc --incremental
  src/test.ts(1,49): error TS2307: Cannot find module 'ava' or its corresponding type declarations.
      at makeError (/data/data/com.termux/files/home/Projects/meronascript/node_modules/execa/lib/error.js:60:11)
      at handlePromise (/data/data/com.termux/files/home/Projects/meronascript/node_modules/execa/index.js:118:26)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Object.compile (file:///data/data/com.termux/files/home/Projects/meronascript/node_modules/@ava/typescript/index.js:100:7)                                                                                                                       at async file:///data/data/com.termux/files/home/Projects/meronascript/node_modules/ava/lib/api.js:243:19                    at async Promise.all (index 0)
      at async Api.run (file:///data/data/com.termux/files/home/Projects/meronascript/node_modules/ava/lib/api.js:242:4)
      at async loadCli (file:///data/data/com.termux/files/home/Projects/meronascript/node_modules/ava/lib/cli.js:489:21)


  ─

error Command failed with exit code 1.
```

**Reply by novemberborn** [reply]
Could you try doing the TypeScript compilation outside of AVA?

**Reply by catink123** [reply]
I tried running the TypeScript compiler directly and it results in the same error:

```
$ yarn tsc
yarn run v1.22.19
$ /data/data/com.termux/files/home/testproj/node_modules/.bin/tsc
src/test.ts:1:49 - error TS2307: Cannot find module 'ava' or its corresponding type declarations.

1 import anyTest, {ExecutionContext, TestFn} from 'ava';
                                                  ~~~~~


Found 1 error in src/test.ts:1

error Command failed with exit code 2.
```

**Reply by novemberborn** [reply]
What does `npx tsc -v` give you?

**Reply by catink123** [reply]
```
$ npx tsc -v
Version 4.7.4
```

**Reply by novemberborn** [reply]
@catink123 OK I think I figured it out. They don't make it easy!

Could you try installing from `avajs/ava#ts-node-resolution-shims`? That's the contents of this PR: https://github.com/avajs/ava/pull/3108

**Reply by catink123** [reply]
Maybe it's a problem with Yarn, but I can't install AVA from `ts-node-resolution-shims` branch:
```
$ yarn add -D avajs/ava#ts-node-resolution-shims
yarn add v1.22.19
[1/4] Resolving packages...
[2/4] Fetching packages...
error https://codeload.github.com/avajs/ava/tar.gz/a76dbc25d291160eee070ad214ddce6a7fbed8a5: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, symlink '' -> '/data/data/com.termux/files/home/.cache/yarn/v6/npm-ava-5.0.0-a76dbc25d291160eee070ad214ddce6a7fbed8a5/node_modules/ava/test/node_modules/ava'"
```

I tried cleaning the Yarn cache, removing node_modules and reinstalling, and even using `network-concurrency 1` option with Yarn and nothing helped. I also tried this on my Windows machine under WSL and it returns the same exact error.

**Reply by novemberborn** [reply]
If you could test with npm I'd appreciate it.

**Reply by catink123** [reply]
Okay, it installed with NPM, but the error with importing AVA is the same as before:
```
$ npm run test

> testproj@0.0.1 test
> ava src/test.ts


  ✘ Internal error
  Error: Command failed with exit code 2: tsc --incremental
  src/test.ts(1,49): error TS2307: Cannot find module 'ava' or its corresponding type declarations.
      at makeError (/data/data/com.termux/files/home/Projects/testproj/node_modules/execa/lib/error.js:60:11)
      at handlePromise (/data/data/com.termux/files/home/Projects/testproj/node_modules/execa/index.js:118:26)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Object.compile (file:///data/data/com.termux/files/home/Projects/testproj/node_modules/@ava/typescript/index.js:100:7)
      at async file:///data/data/com.termux/files/home/Projects/testproj/node_modules/ava/lib/api.js:244:19
      at async Promise.all (index 0)
      at async Api.run (file:///data/data/com.termux/files/home/Projects/testproj/node_modules/ava/lib/api.js:243:4)
      at async loadCli (file:///data/data/com.termux/files/home/Projects/testproj/node_modules/ava/lib/cli.js:489:21)


$ npx tsc

src/test.ts:1:49 - error TS2307: Cannot find module 'ava' or its corresponding type declarations.

1 import anyTest, {ExecutionContext, TestFn} from 'ava';
                                                  ~~~~~

Found 1 error in src/test.ts:1

**Reply by novemberborn** [reply]
To be clear, did you install the branch? Can you pop it into the package.json file to be sure?

**Reply by catink123** [reply]
Yes, it is there:
```json
"devDependencies": {
    "@ava/typescript": "^3.0.1",
    "@types/node": "^18.7.16",
    "ava": "github:avajs/ava#ts-node-resolution-shims",
    "typescript": "^4.7.4"
}
```

**Reply by novemberborn** [reply]
Hmm, could you share your tsconfig? I could trigger the problem and resolve it but there may be something different with your setup.

**Reply by catink123** [reply]
Okay, here's my tsconfig:
```json
{
  "compilerOptions": {
    "module": "ES2020",
    "moduleResolution": "node",
    "rootDir": "src",
    "outDir": "dist",
    "lib": ["es2017", "es2021", "dom"],
    "target": "es2015"
  }
}
```
It actually may be wrong, because I still don't fully understand how to configure the compiler properly.

**Reply by novemberborn** [reply]
I forgot to include the new definition files in the package, but also it didn't work with your combination of `ES2020/node`. Could you try again now?

I think there are five combinations to take care of:

* `commonjs/node`
* `commonjs/node16`
* `node16/node16`
* `es2020/node`
* `es2020/node16`

**Reply by catink123** [reply]
Been busy for a while, but I did what you suggested. I've changed `module` in my tsconfig to `es2020` (in lower-case), added `--experimental-specifier-resolution=node` as a Node argument to AVA configuration and now everything works! I suppose you'll merge the `ts-node-resolution-shims` branch and it'll be added to the 5.0.0 release.

**Reply by novemberborn** [reply]
Does it also work with the original configuration?

Using `node16` for `module` and `moduleResolution` would be the expected configuration for a Node.js project. No need to add any flags within AVA. But you may need to set `"type": "module" in your `package.json` for TypeScript to emit ESM files (and Node.js to treat `.js` files as ESM).

**Reply by catink123** [reply]
It doesn't work with `node16` as `module` with this error:
```
ReferenceError: exports is not defined in ES module scope                       
This file is being treated as an ES module because it has a '.js' file extensi
on and '/data/data/com.termux/files/home/testproj/package.json' con
tains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
``` 
and it doesn't work with the original configuration without the `--experimental-specifier-resolution=node` Node flag, but that's a known problem with TypeScript, so I use this workaround.

Also, I'm developing a library targeting web browsers, so compatibility with Node isn't neccesary (if that's why you were saying about `node16` for `module` for a Node.js project).
